exotations/task_maps/task_map/src/Point2Point.cpp

```diff
@@ -0,0 +1,232 @@
+
+    // This is col(0) of rotation matrix as returned by Kinematics
+
+    cr = std::cos(roll);
+    sr = std::sin(roll);
+
+    cp = std::cos(pitch);
+    sp = std::sin(pitch);
+
+    cy = std::cos(yaw);
+    sy = std::sin(yaw);
+
+    ahat(0) = cr*cp;
+    ahat(1) = sr*cp;
+    ahat(2) = -sp;
+
+    return ahat;
+  }
+
+  Eigen::Vector3d Point2Point::getApproachVector(int i) {
+    return getEffRotation(i).col(0); // approach vector is x axis or exotica/Frame0Alwr_arm_6_link! 
+  }
+
+  Eigen::Matrix3d Point2Point::getEffRotation(int i) {
+    return Eigen::Map<Eigen::Matrix3d>(Kinematics[0].Phi(i).M.data).transpose(); // note, exotica/kdl returns transpose of matrix required
+  }
+
+  Eigen::Matrix3d Point2Point::getApproachVectorJacobian(double roll, double pitch, double yaw) {
+    // Approach vector jacobian (aJ)
+    //
+    // Let Phi = (r, p, y) be rpy euler angles
+    //     ahat = ahat(Phi) be approach vector of eff 
+    //
+    // Then aJ = d ahat(Phi) / d Phi 
+
+    Eigen::Matrix3d J;
+    
+    double cr, sr; // cos/sin(roll)
+    double cp, sp; // cos/sin(roll)
+    double cy, sy; // cos/sin(roll)
+
+    cr = std::cos(roll);
+    sr = std::sin(roll);
+
+    cp = std::cos(pitch);
+    sp = std::sin(pitch);
+
+    cy = std::cos(yaw);
+    sy = std::sin(yaw);
+
+    // Row 1
+    J(0,0) = -sr*cp;
+    J(0,1) = -cr*sp;
+
+    // Row 2
+    J(1,0) = cr*cp;
+    J(1,1) = -sr*sp;
+
+    // Row 3
+    J(2,1) = -cp;
+
+    return J;
+    
+  }
+
+  void Point2Point::update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef phi, Eigen::MatrixXdRef J)
+  {
+
+    // Collect rpy angles
+    double roll, pitch, yaw;
+    Kinematics[0].Phi(0).M.GetRPY(roll, pitch, yaw);
+
+    // Compute eff postion and approach vector
+    Eigen::Vector3d e = getEffPosition(0);
+    Eigen::Vector3d ahat = getApproachVector(0);
+
+    // Setup
+    Eigen::Vector3d eap = e + ahat - point_;
+    Eigen::Vector3d L = point_ - e;
+    Eigen::Vector3d K = ahat.cross(eap);
+    Eigen::Matrix3d J_rpy = getApproachVectorJacobian(roll, pitch, yaw);
+    double LL = L.squaredNorm();
+    double LL2 = LL*LL;
+    double KK = K.squaredNorm();
+    Eigen::MatrixXd J_q = Kinematics[0].J(0).data;
+    
+    // Compute phi
+    phi(0) = KK / LL;
+
+    // Compute J
+    for (int i=0; i<N; i++){
+
+      // Setup
+      Eigen::Vector3d ed = J_q.col(i).topRows(3);
+      Eigen::Vector3d ahatd = J_rpy * J_q.col(i).bottomRows(3);
+      Eigen::Vector3d Kd = ahatd.cross(eap) + ahat.cross(ed + ahatd);
+      Eigen::Vector3d Ld = -ed;
+
+      J(0, i) = 2.0 * (Kd.dot(K)*LL - KK*Ld.dot(L)) / LL2;
+      
+    }
+    
+  }
+  
+  void Point2Point::assignScene(Scene_ptr scene)
+  {
+    scene_ = scene;
+    Initialize();
+  }
+
+  void Point2Point::Initialize() {
+    N = scene_->getSolver().getNumControlledJoints();
+  }
+
+
+  void Point2Point::Instantiate(Point2PointInitializer &init)
+  {
+    point_ = init.Point; 
+  }
+
+  int Point2Point::taskSpaceDim()
+  {
+    return 1;
```

Comment:
That assumes that you can only use one end-effector with this task. If this is desired, throw and error if more are defined (e.g., in `Initialize()`)